README.md
# Woopra Browser Tracker

[**Full Documentation ▸**](https://docs.woopra.com/reference#intro-javscript-sdk)

# Installation

For best results, this should be included at the top of your document, preferrably in the HEAD section,
but definitely before you make any calls to `woopra`. You may use any `woopra` function after this snippet
and it will process all of your commands after the script loads asynchronously.

```html
<script>
  // load woopra tracker
  !function(){var t,o,c,e=window,n=document,r=arguments,a="script",i=["call","cancelAction","config","identify","push","track","trackClick","trackForm","update","visit"],s=function(){var t,o=this,c=function(t){o[t]=function(){return o._e.push([t].concat(Array.prototype.slice.call(arguments,0))),o}};for(o._e=[],t=0;t<i.length;t++)c(i[t])};for(e.__woo=e.__woo||{},t=0;t<r.length;t++)e.__woo[r[t]]=e[r[t]]=e[r[t]]||new s;(o=n.createElement(a)).async=1,o.src="https://static.woopra.com/w.js",(c=n.getElementsByTagName(a)[0]).parentNode.insertBefore(o,c)}("woopra");

  // configure tracker
  woopra.config({
    domain: 'mybusiness.com'
  });

  // track pageview action
  woopra.track();
</script>
```

### Advanced

You can change the tracker's instance name by changing the string in the last line of the snippet `})('woopra');`.
If you need multiple instances just duplicate the snippet with a different instance name.

# Configuration

The Woopra tracker can be customized using the config function. Find below the list of options:

| Option                    | Default                     | Description                                                                                                          |
| ------------------------- | --------------------------- | -------------------------------------------------------------------------------------------------------------------- |
| domain                    | N/A: required               | **REQUIRED** Project Name in Woopra. Must be explicitly set for tracker to work.                                     |
| cookie_name               | `'wooTracker'`              | Name of the cookie used to identify the visitor.                                                                     |
| cookie_domain             | Website domain              | Domain scope of the Woopra cookie.                                                                                   |
| cookie_path               | `/`                         | Directory scope of the Woopra cookie.                                                                                |
| cookie_expire             | 2 years from last action    | Expiration date (Date object) of the Woopra cookie.                                                                  |
| ping                      | `false`                     | Ping woopra servers to ensure that the visitor is still on the webpage.                                              |
| ping_interval             | `12000`                     | Time interval in milliseconds between each ping. (min: `6000`, max: `60000`)                                         |
| download_tracking         | `false`                     | Track downloads on the web page.                                                                                     |
| download_extensions       | `['avi', 'css', 'dmg',...]` | Specifies which file extensions will result in a `downloaded` event to be tracked upon click.                        |
| outgoing_tracking         | `false`                     | Track external links clicks on the web page.                                                                         |
| outgoing_ignore_subdomain | `true`                      | Do not include links to subdomains as outgoing links.                                                                |
| download_pause            | `200`                       | Time in millisecond to pause the browser to ensure that the event is tracked when visitor clicks on a download url.  |
| outgoing_pause            | `200`                       | Time in millisecond to pause the browser to ensure that the event is tracked when visitor clicks on an outgoing url. |
| click_pause               | `250`                       | Time in millisecond to pause the browser to ensure that the event is tracked when visitor clicks on an element url.  |
| form_pause                | `250`                       | Time in millisecond to pause the browser to ensure that the event is tracked when visitor clicks submits a form.     |
| ignore_query_url          | `true`                      | Ignores the query part of the url when the standard pageviews tracking function (track() with no args) is called     |
| map_query_params          | `{}`                        | Object with URL parameter keys mapped to action property names. (e.g. { ref: "campaign_name" })                      |
| hide_campaign             | `false`                     | Enabling this option will remove campaign properties from the URL when they’re captured (using HTML5 `pushState`).   |
| personalization           | `true`                      | Keeps the tracking request connection open to allow personalization features like segments and client triggers       |
| protocol                  | `https://`                  | The protocol used to contact Woopra's servers. `http://` forces unsecure mode and `//` follows the page's protocol   |
| beacons                   | Browser support             | Whether to use `navigator.sendBeacon` instead of `<script>` injection to send actions to Woopra's tracking servers   |

The `config()` function supports key/value singleton argument:

```javascript
woopra.config('cookie_name', 'my_business_cookie');
```

Or an object of keys and values:

```javascript
woopra.config({
  download_tracking: false,
  outgoing_tracking: false,
  ping_interval: 30000
});
```

# Identifying Customers

In order to identify a customer, you need to send their email address to Woopra as a custom visitor property. It should
be added before the `track()` function and formatted in the following manner:

```javascript
// Identify customer before the track event
woopra.identify({
  email: 'johndoe@mybusiness.com',
  name: 'John Doe',
  company: 'My Business'
});
```

Standard attributes which will be displayed in the Woopra profiles include:

- `email` - Which displays the visitor's email address and it will be used as a unique identifier instead of cookies.
- `name` - Which displays the visitor's full name
- `company` - Which displays the company name or account of your customer

Note that if you wish to identify a visitor without sending a tracking event, you can call the function `push()`.
`push()` also accepts a callback function as a parameter.

```javascript
woopra.identify('email', 'johndoe@mybusiness.com').push();
```

# Tracking Actions

Woopra also allows you to track Custom Actions in addition to simple pageviews. Let’s say you are running a website
where people can signup. You can track these actions using Woopra’s Custom Events.

The `track()` function will track the pageview of the current page if no arguments are provided:

### Arguments

- `eventName` - name of the custom event to track
- `properties` - An object containing any properties to send with the custom event
- `callback` - A callback function on success

```javascript
woopra.track();

// The line above is equivalent to
woopra.track('pv', {
  url: window.location.pathname + window.location.search,
  title: document.title
});
```

To track custom actions, you can define the action name string and the properties associated with that action object:

```javascript
woopra.track(
  'signup',
  {
    company: 'My Business',
    username: 'johndoe',
    plan: 'Gold'
  },
  function (action) {
    // track request successfully sent to Woopra
  }
);
```

The code above will track a custom event titled `signup` and provides some more information like the username and company of the account created. Just imagine all of the custom events that can be tracked on your website: payment, comment, logout, email, etc…

What’s even more important about custom events is that you can always run custom reports about the data you pass to Woopra, so for the example given above, you could get the number of signup by company.

When you track custom events, remember to update your schema on Woopra. That will help the rest your team to identify the events being tracked and Woopra will automatically build reports out of the box based on the event properties.

Below is another example to track when people click on the play button with id="play-button":

```javascript
document.getElementById('play-button').onclick = function() {
  woopra.track('play', {
    artist: 'Dave Brubeck',
    song: 'Take Five',
    genre: 'Jazz'
  });
};
```

# Browser Support

- Chrome 15+
- Firefox 6+
- Safari 5.1+
- Edge
- Internet Explorer 9+


CHANGELOG.md
# v5.3.0 (2021-09-15)
- `woopra.track`:
  - Updated API to accept 3rd parameter as an object instead of a callback function
  - `onError` callback option
  - `onBeforeSend` callback option, receives a `WoopraAction` object as an argument
  - `onSuccess` callback option, receives a `WoopraAction` object as an argument
  - `timeout` option (`Number`, default `woopra.config('idle_timeout')`)
  - `queue` option (`Boolean`)
    - When true, forces the action to be queued and sent as a beacon
    - Uses `navigator.sendBeacon` instead of `<script>` tag injection to make the tracking request
    - Queued actions are flushed and sent to woopra's tracking servers every time the user closes the tab/window or switches away from the tab, navigates away from the page, `woopra.push()` is called, an action has been cancelled or a file download/outgoing link/form track/click track event is captured.
  - `lifecycle` option (`String: "page" | "action"`)
    - Defaults to `page` for `pv`, otherwise `action`
    - Actions with `lifecycle` of `page` will update durations through the lifecycle of the page
    - Actions with `lifecycle` of `page` will be updated with the % of the page the user has seen upon scrolling
  - `retrack` option (`Boolean`, default `false`)
    - Actions with `lifecycle` of `page` will be tracked again if the user is inactive longer than `timeout`. The action has the same properties with an extra `returning` property with the value of `true`
  - Errors thrown in callbacks will be caught and logged via `console.error`
- `woopra.update`:
  - New API to update existing actions using an ID of a `WoopraAction`
  - `woopra.update(id, { updated: 'property' });`
- `woopra.cancelAction`:
  - New API to cancel actions with `lifecycle` of `page` from further updating themselves using an ID of a `WoopraAction`
  - `woopra.cancelAction(id);`
- `WoopraAction`:
  - Object returned from `onSuccess` and `onBeforeSend` callbacks
  - `action.id`
  - `action.update({ updated: 'property' })`
    - Equivalent to `woopra.update(action.id, { updated: 'property' })`
  - `action.cancel()`
    - Equivalent to `woopra.cancelAction(action.id)`
- `woopra.config`
  - `beacons` (`Boolean`, defaults to browser support)
    - Enable/disable usage of beacons
  - `download_extensions` (`Array`)
    - List of extensions (without '.') to treat as downloadable files
  - `click_pause` (`Number`, default `250`)
    - Time in millisecond to pause the browser to ensure that the event is tracked when visitor clicks on an element decorated with `woopra.trackClick`. This value is ignored when `beacons` are enabled.
  - `form_pause` (`Number`, default `250`)
    - Time in millisecond to pause the browser to ensure that the event is tracked when visitor clicks on an element decorated with `woopra.trackForm`. This value is ignored when `beacons` are enabled.
- Replace utility functions with cherry-picked `lodash-es`
- Include https://github.com/GoogleChromeLabs/page-lifecycle inline to handle consistent browser behavior


woopra_v3.js
(function(window, document) {
    "use strict";

    var WoopraScript = function(_src, _hook, _async) {
        this.scriptObject = false;
        this.src = _src;
        this.hook = _hook;
        this.async = _async;
    };

    WoopraScript.prototype.clear = function() {
        this.scriptObject.parentNode.removeChild(this.scriptObject);
    };

    WoopraScript.prototype.load = function() {
        var script = document.createElement('script'),
            ssc,
            _hook = this.hook;

        script.src = this.src;
        script.async = this.async;

        if (_hook) {
            if (typeof script.onreadystatechange !== 'undefined') {
                script.onreadystatechange = function() {
                    if (this.readyState === 'complete'|| this.readyState === 'loaded') {
                        window.setTimeout(_hook, 400);
                    }
                };
            }
            else {
                script.onload = function() {
                    window.setTimeout(function() {
                        _hook.apply();
                    }, 400);
                };
            }
        }

        ssc = document.getElementsByTagName('script')[0];
        ssc.parentNode.insertBefore(script, ssc);
    };

    var WoopraEvent = function(name, ce, cv, file) {
        this.name = name || 'unknown';
        this.ce = ce || {};
        this.cv = cv || {};
        this.file = file || 'ce';
        this.requestString = '';
        this.attachCampaignData();
    };

    WoopraEvent.prototype.attachCampaignData = function() {
        var vars = this.getUrlVars(),
            i,
            key,
            value,
            campaignKeys = ['source', 'medium', 'content', 'campaign', 'term'];

        for (i = 0; i < campaignKeys.length; i++) {
            key = campaignKeys[i];
            value = vars['utm_' + key] || vars['woo_' + key];

            if (typeof value !== 'undefined') {
                this.ce['woo_' + key] = value;
            }
        }
    };

    WoopraEvent.prototype.getUrlVars = function() {
        var vars = {};

        window.location.href.replace(/[?&]+([^=&]+)=([^&]*)/gi, function(m, key, value) {
            vars[key] = value;
        });
        return vars;
    };

    WoopraEvent.prototype.addProperty = function(key, value) {
        this.ce[key] = value;
    };

    WoopraEvent.prototype.serialize = function(v, prefix) {
        var i,
            itemKey;

        if (this.requestString.length > 4000 ||
            typeof v === 'undefined' ||
            typeof v === 'function') {
            return;
        }

        if (typeof v === 'string' ||
            typeof v === 'number' ||
            typeof v === 'boolean') {
            this.requestString += '&' + encodeURIComponent(prefix) +
                                  '=' + encodeURIComponent(v);
            return;
        }

        if (v instanceof Array) {
            for (i=0; i < v.length; i++) {
                this.serialize(v[i], prefix + '[' + i + ']');
            }
            return;
        }

        for (itemKey in v) {
            if (v.hasOwnProperty(itemKey)) {
                this.serialize(v[itemKey], prefix + '.' + itemKey);
            }
        }
    };

    WoopraEvent.prototype.fire = function(tracker) {
        var t = tracker,
            key,
            rd = t.getRequestData() || {};

        t.ensure();

        this.addProperty('name', this.name);
        this.requestString = '';

        for (key in rd) {
            if (rd.hasOwnProperty(key)){
                this.serialize(rd[key], key);
            }
        }
        for (key in this.cv) {
            if (this.cv.hasOwnProperty(key)){
                this.serialize(this.cv[key], 'cv_'+key);
            }
        }

        for (key in this.ce){
            if (this.ce.hasOwnProperty(key)){
                this.serialize(this.ce[key], 'ce_'+key);
            }
        }

        new WoopraScript(t.getEndpoint(this.file)+'?ra='+t.randomstring()+this.requestString, function(){}, true ).load();

    };

    var WoopraTracker = function() {
        this.chat = false;
        this.alias = '';
        this.vs = 0;
        this.props = {};
        this.cv = {};
        this.cs = {};
        this.pint = false;
        this.ctr = 0;
        this.version = 20;
        this.last_activity = new Date();
        this.idle = 0;
        this.inited = 0;
        this._loaded  =  false;
        this.initialize.apply(this, arguments);
    };

    WoopraTracker.prototype.initialize = function() {
        var i,
            a,
            _wpt = window._wpt,
            _this = this;

        this._setOptions();

        this._processQueue();
        this._loaded = true;
        if (window.woopraReady !== 'undefined') {
            window.woopraReady(this);
        }

    };

    /**
     * Processes the tracker queue in case user tries to push events
     * before tracker is ready.
     */
    WoopraTracker.prototype._processQueue = function() {
        var i, a,
            _wpt = window._wpt;

        if (_wpt && _wpt._e) {
            for (i = 0; i < _wpt._e.length; i++) {
                a = _wpt._e[i];
                this[_wpt._e[i][0]].apply(this, Array.prototype.slice.call(a, 1));
            }
        }
        window._wpt = this;
    };

    /**
     * Sets the initial options
     */
    WoopraTracker.prototype._setOptions = function() {
        var t = this,
            exp = new Date();

        t.option('domain', window.location.hostname);
        t.option('cookie_name', 'wooTracker');
        t.option('cookie_domain', window.location.hostname);
        t.option('cookie_path', '/');
        exp.setDate(exp.getDate()+365);
        t.option('cookie_expire', exp);
        t.option('ping', true);
        t.option('ping_interval', 12000);
        t.option('idle_timeout', 300000);
        t.option('download_pause', 200);
        t.option('outgoing_pause', 400);
        t.option('download_tracking', true);
        t.option('outgoing_tracking', true);
    };

    WoopraTracker.prototype.ensure = function() {
        var t = this,
            name,
            _c;

        if (t.inited === 0) {
            t.inited = 1;

            name = t.option('cookie_name');
            _c = t.readcookie(name);
            if (!_c || _c.length <= 0) {
                _c = t.randomstring();
            }
            t.createcookie(name, _c);

            window.setInterval(function() {
                t.next();
            }, 1000);

            if (typeof document.attachEvent !== 'undefined') {
                document.attachEvent("onmousedown", function() {
                    t.clicked.apply(t, arguments);
                });
                document.attachEvent("onmousemove", function() {
                    t.moved.apply(t, arguments);
                });
                document.attachEvent("onkeydown", function() {
                    t.typed.apply(t, arguments);
                });
            } else {
                document.addEventListener("mousedown", function() {
                    t.clicked.apply(t, arguments);
                }, false);
                document.addEventListener("mousemove", function() {
                    t.moved.apply(t, arguments);
                }, false);
                document.addEventListener("keydown", function() {
                    t.typed.apply(t, arguments);
                }, false);
            }
        }
    };

    WoopraTracker.prototype.getEndpoint = function(file) {
        return window.location.protocol + '//www.woopra.com/track/' + file + '/';
    };

    WoopraTracker.prototype.sleep = function(millis) {
        var date = new Date(),
            curDate = new Date();

        while (curDate-date < millis) {
            curDate = new Date();
        }
    };

    WoopraTracker.prototype.randomstring = function() {
        var chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz',
            i,
            rnum,
            s = '';

        for (i = 0; i < 12; i++) {
            rnum = Math.floor(Math.random() * chars.length);
            s += chars.substring(rnum, rnum + 1);
        }

        return s;
    };

    WoopraTracker.prototype.readcookie = function(k) {
        var c = document.cookie + '',
            ind,
            ind1;

        ind = c.indexOf(k);
        if (ind === -1 || k === '') {
            return '';
        }

        ind1 = c.indexOf(';', ind);
        if (ind1 === -1) {
            ind1 = c.length;
        }

        return window.unescape(c.substring(ind + k.length + 1, ind1));
    };

    WoopraTracker.prototype.createcookie = function(k, v) {
        var t = this,
            cookieval;

        cookieval = k + '=' + v + '; ' + 'expires=' + t.option('cookie_expire') + ';' + 'path='+t.option('cookie_path')+';domain=.'+t.option('cookie_domain');
        document.cookie = cookieval;
    };

    WoopraTracker.prototype.getRequestData = function(){
        var t = this,
            r = {};

        r.alias = t.option('domain');
        r.cookie = t.readcookie(t.option('cookie_name'));
        r.meta = t.readcookie('wooMeta') || '';
        r.screen = window.screen.width + 'x' + window.screen.height;
        r.language = (window.navigator.browserLanguage || window.navigator.language || "");
        r.referer = document.referrer;
        r.idle = parseInt(t.idle/1000, 10) + '';

        if (t.vs === 2) {
            r.vs = 'w';
            t.vs = 0;
        } else {
            if (t.idle == 0) {
                r.vs = 'r';
            } else{
                r.vs = 'i';
            }
        }

        return r;
    };

    /**
     * Sends a pageview event to the server
     *
     * Alias for `track('pv', options)`
     */
    WoopraTracker.prototype.pageview = function(options) {
        var _options = options || {},
            e;

        if (typeof _options.visitor !== 'undefined') {
            this.cv = _options.visitor;
            delete _options.visitor;
        }

        e = new WoopraEvent('pv', _options, this.cv, 'visit');
        e.fire(this);
    };

    /**
     * Sends a custom event to the server
     *
     * If you pass a `visitor` object to options, then the custom visitor
     * data in `visitor` will be sent along with the custom event
     *
     * This is equivalent to calling `_wpt.visitor(data)` before
     * calling `track()`
     */
    WoopraTracker.prototype.track = function(name, options) {
        var _options = options || {},
            e;

        if (typeof _options.visitor !== 'undefined') {
            this.cv = _options.visitor;
            delete _options.visitor;
        }

        e = new WoopraEvent(name, _options, this.cv, 'ce');
        e.fire(this);
    };

    /**
     * Attach custom visitor data and then sends data to server
     * We use `email` as our unique id, so make sure you use the email as
     */
    WoopraTracker.prototype.identify = function(email, properties) {
        var e;
        if (typeof properties !== 'undefined') {
            this.cv = properties;
        }
        if (typeof email !== 'undefined' && email !== '') {
            this.visitor('email', email);
        }

        e = new WoopraEvent('identify', {}, this.cv, 'identify');
        e.fire(this);
    };

    /**
     * Gets/sets tracker options
     */
    WoopraTracker.prototype.option = function(k, v) {
        if (typeof v === 'undefined') {
            return this.props[k];
        }

        this.props[k] = v;
        return v;
    };

    /**
     * Attach custom visitor data without sending the server an event.
     * Use identify to sync the visitor data to the server
     */
    WoopraTracker.prototype.visitor = function(name, value) {
        if (typeof name === 'string' && typeof value !== 'undefined') {
            this.cv[name] = value;
        }
        else if (typeof name === 'object') {
            this.cv = name;
        }
    };

    /**
     * Use to attach custom visit data that doesn't stick to visitor
     * ** Not in use yet
     */
    WoopraTracker.prototype.visit = function(name, value) {
        if (typeof name === 'string' && typeof value !== 'undefined') {
            this.cs[name] = value;
        }
        else if (typeof name === 'object') {
            this.cs = name;
        }
    };

    /** Compatibility with old tracker methods **/
    /**
     * Shortcut to set domain
     */
    WoopraTracker.prototype.setDomain = function(domain) {
        this.option('domain', domain);
        this.option('cookie_domain', domain);
    };

    WoopraTracker.prototype.setIdleTimeout = function(timeout) {
        this.option('idle_timeout', timeout);
    };

    // XXX
    WoopraTracker.prototype.pingServer = function() {
        var e = new WoopraEvent('x', {}, this.cv, 'ping');
        e.fire(this);
    };

    WoopraTracker.prototype.typed = function(e) {
        this.vs = 2;
    };

    WoopraTracker.prototype.clicked=function(e) {
        var t = this;

        t.moved();

        var cElem = e.srcElement || e.target;
        while (typeof cElem !== 'undefined' && cElem !== null){
            if (cElem.tagNam === 'A') {
                break;
            }
            cElem = cElem.parentNode;
        }

        if (typeof cElem !== 'undefined' && cElem !== null){
            var link=cElem;
            var _download = link.pathname.match(/(?:doc|dmg|eps|jpg|jpeg|png|svg|xls|ppt|pdf|xls|zip|txt|vsd|vxd|js|css|rar|exe|wma|mov|avi|wmv|mp3|mp4|m4v)($|\&)/);
            var ev=false;
            if(t.option('download_tracking')){
                if(_download && (link.href.toString().indexOf('woopra-ns.com')<0)){
                    ev=new WoopraEvent('download', {url:link.href});
                    ev.addProperty('url',link.href);
                    ev.fire(this);
                    t.sleep(t.option('download_pause'));
                }
            }
            if (t.option('outgoing_tracking')) {
                if (!_download &&
                    link.hostname !== window.location.host &&
                    link.hostname.indexOf('javascript') === -1 &&
                    link.hostname !== '') {
                    ev = new WoopraEvent('outgoing', {url:link.href});
                    ev.fire(this);
                    t.sleep(t.option('outgoing_pause'));
                }
            }
        }
    };

    WoopraTracker.prototype.moved = function() {
        var t = this;
        t.last_activity = new Date();
        t.idle = 0;
    };

    WoopraTracker.prototype.next = function(){
        var now,
            t = this;

        //clocks every 1 second
        t.ctr++;

        if (t.ctr * 1000 > t.option('ping_interval')) {
            t.ctr = 0;
            if (t.option('ping') && t.idle < t.option('idle_timeout')) {
                t.pingServer();
            }
        }

        now = new Date();
        if (now-t.last_activity > 10000) {
            t.idle = now - t.last_activity;
        }
    };

    var woopraTracker = new WoopraTracker();

    if (typeof exports !== 'undefined') {
        exports.WoopraTracker = WoopraTracker;
        exports.WoopraEvent = WoopraEvent;
    }
}(window, document));


